include/threads.h

```diff
@@ -26,9 +26,24 @@
 #ifndef _THREADS_H_
 #define _THREADS_H_
 
-void logging_thread_cleanup(void *arg);
+#include <pthread.h>
+#include <unistd.h>
 
 void *logging_thread(void *arg);
 void *reading_thread(void *arg);
 
+// see https://blog.memzero.de/pthread-cancel-noexcept/
+
+#define CANCELLABLE(operation)                                                                     \
```

Comment:
still two worries:
a) https://hownot2code.wordpress.com/2016/12/05/do-while-0-in-macros/ I wonder whether it would be better instead of the {} to do a do{operation;}while(0)

b) as "operation" could easily contain more complex expressions  it might be better to use:
#define CANCELLABLE(...) and then later on do{ __VA_ARGS__; } while(0)

Reply:
nice trick!